lib/access-context.js

```diff
@@ -86,8 +86,9 @@ function AccessContext(context) {
 
   var token = this.accessToken || {};
 
-  if (token.userId != null) {
-    this.addPrincipal(Principal.USER, token.userId);
+  if (token.userId) {
+    const userPrincipalType = token.principalType || Principal.USER;
+    this.addPrincipal(userPrincipalType, token.userId);
```

Comment:
I find this change suspicions - I would expect that `token.principalType` was already added on line 84 above.

Having said that, this change does fix certain bugs like e.g. #3829, so it's not without merit. I need to build a better understanding of this part of our codebase to be able to comprehend impacts of this change in a wider context.

Reply:
Never mind, the line 84 was added back in 2013, before support for multiple user models were introduced. I suspect that code path was actually never used.

Reply:
@bajtos You are right about line 84, but this block of code is neglecting the principal type associated with the accessToken,

https://github.com/strongloop/loopback/blob/6ddf268cb626521971f5e5b63382bf08af78285a/lib/access-context.js#L47-L85

I think we will need to correct the whole block for a holistic solution

Reply:
My initial thought is, if line 84 was supposed to set the correct principal type then line 47 is dumping the empty object in some corner case.